kitodo-mediaserver-core/src/main/java/org/kitodo/mediaserver/core/services/ActionService.java

```diff
@@ -0,0 +1,183 @@
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * LICENSE file that was distributed with this source code.
+ */
+
+package org.kitodo.mediaserver.core.services;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import org.kitodo.mediaserver.core.api.IAction;
+import org.kitodo.mediaserver.core.db.entities.ActionData;
+import org.kitodo.mediaserver.core.db.entities.Work;
+import org.kitodo.mediaserver.core.db.repositories.ActionRepository;
+import org.kitodo.mediaserver.core.exceptions.ActionServiceException;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+/**
+ * Run Actions.
+ */
+@Service
+public class ActionService {
+
+    private ApplicationContext applicationContext;
+
+    private ActionRepository actionRepository;
+
+    @Autowired
+    public void setApplicationContext(ApplicationContext applicationContext) {
+        this.applicationContext = applicationContext;
+    }
+
+    @Autowired
+    public void setActionRepository(ActionRepository actionRepository) {
+        this.actionRepository = actionRepository;
+    }
+
+    private ActionService() {}
+
+    /**
+     * Request a action and make it persistent.
+     * @param work the Work
+     * @param actionName action bean name to be used
+     * @param parameter parameter list
+     */
+    public void request(Work work, String actionName, Map<String, String> parameter) throws ActionServiceException {
+
+        if (work == null) {
+            throw new IllegalArgumentException("work must not be null.");
+        }
+        if (!StringUtils.hasText(actionName)) {
+            throw new IllegalArgumentException("actionName must be an action Bean name.");
+        }
+
+        // if this exact action is already requested and not finished, break
+        ActionData actionData = getUnfinishedAction(work, actionName, parameter);
+        if (actionData != null) {
+            throw new ActionServiceException("The requested actionName='" + actionName + "' for workId='" + work.getId() + "' has already been requested.");
+        }
+
+        actionData = new ActionData(work, actionName, parameter);
+        actionData.setRequestTime(Instant.now());
+        actionRepository.save(actionData);
+    }
+
+    /**
+     * Run a saved ActionData.
+     * @param work the Work
+     * @param actionName action bean name to be used
+     * @param parameter parameter list
+     * @return the result of the Action
+     * @throws Exception on Action errors
+     */
+    public Object performRequested(Work work, String actionName, Map<String, String> parameter) throws Exception {
```

Comment:
This method can only be performed, when you already know the action in question. For the case, that one process request an action and another looks for actions to perform at some later time, that will not be the case. Then you would have a DB-query getting ActionData objects to perform and would expect a moethod performRequested(AcionData actionData). Perhaps we should offer both methods, the now implemented one calling the second after getting the ActionData object.